Calculator Part 1

yahelroro0196 opened the pull request.

yahelroro0196 committed:
initial project creation

yahelroro0196 committed:
initial design for project.

yahelroro0196 committed:
added main java file

Commit:
updated Design graph

Commit:
updated design graph

Commit:
1: built config loader and parser
2: built menu call from main
3: built first menu loop logic
4: built input builder and validator

Commit:
added valid operators config file

Commit:
1: changed from navigable map to ArrayList<Pair<>> to support repeating numbers.
2: added calculation logic from infix to postfix

Commit:
1: changed from navigable map to ArrayList<Pair<>> to support repeating numbers.
2: added calculation logic from infix to postfix

yahelroro0196 committed:
1: Implemented complete shunting yard algorithm infix->postfix equation
2: Implemented postfix->result
3: Clean code with constants and short functions
4: Implemented abstract operator.java class + all the different operators
5: created OperatorFactory.
6: changed Pair attributes to better reflect purpose
7: Constants class

yahelroro0196 committed:
1: separated infix to postfix logic from postfix to result logic to same package but different classes.
2: more constant separation
3: added exit function by entering an empty input twice.

yahelroro0196 committed:
1: added more functions to separate and make the code cleaned and more understandable
2: added an InputFlow package to separate input parsing to it's own area
3: move config loading to CalculatorGui to clean menu.

yahelroro0196 committed:
deleted useless space in Main.java

Commit:
removed unused imports in main.

Commit:
updated README.md file.

ik8 commented on .gitignore:
```diff
@@ -21,3 +21,7 @@
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+
+.idea
```
Add '/' at the end. It's a directory

ik8 commented on out/production/calculator/com/company/Config/ValidOperators.properties:
```diff
@@ -0,0 +1 @@
+validOperators=+,-
```
why is there an out directory? it should not be in the repo

ik8 commented on src/com/company/Calculator/CalculationSteps/InfixToPostfix.java:
```diff
@@ -0,0 +1,63 @@
+package com.company.Calculator.CalculationSteps;
+
+import com.company.Calculator.Operators.Operator;
```
You should select a good package name.

ik8 commented on src/com/company/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,52 @@
+        return VALID_INPUT;
+    }
+
+    public static String receiveInput() {
```
Is it CalculatorLogic or CalculatorGUI?

ik8 commented on src/com/company/Calculator/CalculatorGui.java:
```diff
@@ -0,0 +1,78 @@
+
+public class CalculatorGui {
+
+    public static final String CONFIG_PATH = "src\\com\\company\\config\\ValidOperators.properties";
```
When you compile this to a jar it wont work.
you should use as relative route

yahelroro0196 committed:
Delete ValidOperators.properties

yahelroro0196 committed:
new line at end of file

yahelroro0196 committed:
new line at end of file

yahelroro0196 committed:
new line at end of file

ik8 commented on src/com/company/Calculator/CalculatorGui.java:
```diff
@@ -0,0 +1,78 @@
+    private static void displayResult(ArrayList<Pair<String, String>> equation) {
+        try {
+            System.out.printf("Equation result is: %s" + NEW_LINE, calculate(equation));
+        } catch (Exception invalidEquation) {
```
You can't catch exception, you will not know what to do in each case(many exceptions can happen)

ik8 commented on src/com/company/Calculator/CalculatorGui.java:
```diff
@@ -0,0 +1,78 @@
+    private static final String NEW_LINE = "\n";
+    private static final ArrayList<String> VALID_OPERATORS = ConfigLoader.loadConfig(CONFIG_PATH);
+
+    public static void menu() {
```
![image](https://user-images.githubusercontent.com/17217900/79755584-08eb8180-8322-11ea-9961-4850a8de03a4.png)

ik8 replied:
Why are you not checking if the equation is ok? I used the valid operators, what is wrong? fix it.

ik8 commented on src/com/company/Calculator/CalculatorGui.java:
```diff
@@ -0,0 +1,78 @@
+        return equation;
+    }
+
+    public static void displayInvalidInput() {
```
why not create a logger component? what will happen if I want to change to log4j and not use prints? you need to go over all the code?

ik8 commented on src/com/company/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,52 @@
+                                        ArrayList<Pair<String, String>> equation) {
+        buildInput(equation, previousInput, input, validOperators);
+        Pair<String, String> lastEntry = equation.get(equation.size() - 1);
+        switch (lastEntry.getType()) {
```
where is the validation for the whole equation?

ik8 commented on src/com/company/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,52 @@
+        return new ArrayList<>();
+    }
+
+    public static boolean isExitCalculator(String previousResult) {
```
redundant function, write it in the while

ik8 commented on src/com/company/Calculator/InputFlow/InputValidator.java:
```diff
@@ -0,0 +1,48 @@
+
+    public static final String EMPTY_INPUT = "";
+
+    public static String isValidInput(String previousInput, String input, ArrayList<String> validOperators) {
```
why you are not using the Constants class?

ik8 commented on src/com/company/Calculator/InputFlow/InputValidator.java:
```diff
@@ -0,0 +1,48 @@
+
+public class InputValidator {
+
+    public static final String EMPTY_INPUT = "";
```
config

ik8 commented on src/com/company/Calculator/InputFlow/InputValidator.java:
```diff
@@ -0,0 +1,48 @@
+        return true;
+    }
+
+    private static boolean isOperator(String input, ArrayList<String> validOperators) {
```
you have a lot of one line function you use only in one place, redundant!

ik8 commented on src/com/company/Calculator/Operators/OperatorFactory.java:
```diff
@@ -0,0 +1,18 @@
+            case "/":
+                return new Divide();
+            default:
+                return new Power();
```
power is always a default option?

ik8 commented on src/com/company/Config/ConfigLoader.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static ArrayList<String> loadConfig(String CONFIG_PATH) {
+        Properties prop = new Properties();
+        try (FileReader configFile = new FileReader(CONFIG_PATH)) {
```
use StreamReader, in compiled jar it will not work

ik8 commented on src/com/company/Config/ConfigLoader.java:
```diff
@@ -0,0 +1,23 @@
+            String validOperatorsText = prop.getProperty(VALID_OPERATORS);
+            return ConfigParser.parseValidOperatorsString(validOperatorsText);
+        } catch (IOException e) {
+            e.printStackTrace();
```
do not print the stack, print an indicative message

ik8 commented on src/com/company/Constants.java:
```diff
@@ -0,0 +1,15 @@
+package com.company;
+
+public final class Constants {
```
Turn it into an enum

ik8 commented on src/com/company/Pair.java:
```diff
@@ -0,0 +1,28 @@
+package com.company;
+
+public class Pair<L, R> {
```
nice, but as you see it exists:
https://www.baeldung.com/java-pairs

ik8 commented on src/com/company/Calculator/CalculationSteps/InfixToPostfix.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private static boolean isHigherPrecedence(String operatorText, String subOperatorText) {
+        Operator operator = new OperatorFactory().factory(operatorText);
```
create one object and use the function on him

Commit:
1: updated package name
2: separated GUI and Logic further
3: Added Logging and Log4j component
4: Changed equation array to <String, Type> with ENUM support
5: added 'Type' enum
6: added operator delimiter constant
7: separated constants from enums
8: cleaned Operator
9: changed operator factory to have default and unreachable case
10: added maven
11: added detection of incorrect equation
12: added log4j config file
13: added log4j log file (log.out)

Commit:
deleted useless lines in pom.xml

Commit:
Merge remote-tracking branch 'origin/calculator_part_1' into calculator_part_1

yahelroro0196 committed:
cleaned properties and gitignore

yahelroro0196 committed:
cleaned properties and gitignore

yahelroro0196 committed:
added division by zero handling.

yahelroro0196 committed:
added support to brackets and fixed InfixToPostfix parsing.

yahelroro0196 committed:
cleaned directory

Commit:
updated config to match part 1.

ik8 commented on src/main/java/ShuntingYardCalculator/Pair.java:
```diff
@@ -0,0 +1,28 @@
+package ShuntingYardCalculator;
```
why do you still have it?

ik8 commented on src/main/java/ShuntingYardCalculator/Logging/log.out:
```diff
@@ -0,0 +1,237 @@
+2020-04-20 20:07:04 INFO  Log4j:14 - Welcome to the calculator!
```
should not be here

ik8 commented on src/main/java/ShuntingYardCalculator/Logging/Log4j.java:
```diff
@@ -0,0 +1,45 @@
+package ShuntingYardCalculator.Logging;
```
My meaning was if you use the function only once do not create a function for it
leave 2/3 function that know how to print logs in various situations

ik8 commented on src/main/java/ShuntingYardCalculator/Constants.java:
```diff
@@ -0,0 +1,9 @@
+package ShuntingYardCalculator;
```
Not good for config class.
You can use properties file or if you use it one location and related only to the gui input, just type it in the start of the classs

ik8 commented on src/main/java/ShuntingYardCalculator/Config/ConfigLoader.java:
```diff
@@ -0,0 +1,26 @@
+
+    public static ArrayList<String> loadConfig(String CONFIG_PATH) {
+        Properties prop = new Properties();
+        try (InputStream inputStream = new FileInputStream(CONFIG_PATH)) {
```
**StreamReader**

ik8 commented on pom.xml:
```diff
@@ -0,0 +1,29 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>groupId</groupId>
```
should be unique. do you know why?

Commit:
1: changed Pair to javafx.util.Pair
2: got rid of  Constants.java
3: changed ConfigLoader to StreamReader

Commit:
added equation parsing and calculation unit testing!

Commit:
spacing

Commit:
added example of edge case test in separate class.

Commit:
deleted unused variable

Commit:
added cleaner exception handling to zero division, empty brackets  and more.

Commit:
added exception handling of broken bracket form.

Commit:
added switch case 'break;'

yahelroro0196 committed:
1: added cleaner unit testing code
2: divided tests between classes
3: added easier input, a new parser to parse a string equation.
4: better variable names
5: created a class for each edge case

yahelroro0196 committed:
switched to expression lambda.

yahelroro0196 committed:
1: added more unit testing cases
2: cleaned useless classes and instead added JUnit annotations to the tests
3: fixed test parser to work with multiple digit numbers
4: changed string equation input to be with spaces

yahelroro0196 committed:
1. got rid of overlaying useless package
2. added more edge cases
3. changed variable namings to be clearer with UPPERCASE

yahelroro0196 committed:
1. added config class to hold constants and paths.
2. changed some namings to camelCase
3. updated the unit tests with those variables from the config
class.

yahelroro0196 committed:
Added a more generic parsing mechanism.
